server/paymentProviders/braintree/helpers.ts

```diff
@@ -0,0 +1,113 @@
+import braintree from 'braintree';
+import { isEmpty, isNil, omitBy } from 'lodash';
+
+import { PAYMENT_METHOD_SERVICE } from '../../constants/paymentMethods';
+import { TransactionTypes } from '../../constants/transactions';
+import { convertToCurrency } from '../../lib/currency';
+import { floatAmountToCents } from '../../lib/math';
+import { getHostFee, getPlatformFee } from '../../lib/payments';
+import models, { Op } from '../../models';
+
+const extractFees = (braintreeTransaction): number => {
+  if (!isNil(braintreeTransaction['paypal']?.transactionFeeAmount)) {
+    return floatAmountToCents(braintreeTransaction['paypal'].transactionFeeAmount);
+  } else {
+    // TODO(Braintree): find a way to retrieve this info for non-paypal transactions
+    return 0;
+  }
+};
+
+export const cleanBraintreeTransactionForData = (
+  braintreeTransaction: braintree.Transaction,
+): Record<string, unknown> => {
+  const excludedFields = [
+    'samsungPayCard',
+    'visaCheckoutCard',
+    'androidPayCard',
+    'applePayCard',
+    'localPayment',
+    'merchantAddress',
+    'disbursementDetails',
+    'billing',
+  ];
+
+  return omitBy(braintreeTransaction, (value, key) => {
+    return excludedFields.includes(key) || (typeof value === 'object' && isEmpty(value));
+  });
+};
+
+export class BraintreeTransactionAlreadyExistsError extends Error {
+  braintreeTransaction: braintree.Transaction;
+  dbTransaction: typeof models.Transaction;
+
+  constructor(braintreeTransaction: braintree.Transaction, dbTransaction: typeof models.Transaction) {
+    super(`Braintree transactions for ${braintreeTransaction.id} already exist, skipping.`);
+    this.dbTransaction = dbTransaction;
+    this.braintreeTransaction = braintreeTransaction;
+  }
+}
+
+export const createTransactionsPairFromBraintreeTransaction = async (
+  order: typeof models.Order,
+  braintreeTransaction: braintree.Transaction,
+): Promise<typeof models.Transaction> => {
+  const host = await order.collective.getHostCollective();
+
+  // Make sure the transaction is not already recorded
+  const existingTransactions = await models.Transaction.findAll({
+    where: {
+      HostCollectiveId: host.id,
+      data: { braintreeTransaction: { id: braintreeTransaction.id } }, // TODO(Braintree): Add index on this field
```

Comment:
Yep. I'm a bit afraid of the complexity of having a conditional index and maintaining it.
Not for now, but IMO having a column to store these keys would be simpler.

Reply:
Conditional indexes are actually pretty simple to create, and a dedicated column would require an index too if we want to match the performances. In fact, with a lot of the values being `NULL`, it would be a good thing to use a conditional index even if we go with a dedicated column - there's no need to index a `providerTransactionId` for transactions paid with the collective balance. And ideally, it's more performant if you don't mix up Stripe transactions and Braintree transactions in the same index. 

Regarding the maintenance, there's normally not much to do once the index is in place, so I don't see where the complexity is coming from.

In this specific case, creating the index is just a matter of running this query:
```sql
CREATE INDEX transactions_braintree_id
ON "Transactions" (data -> 'braintreeTransaction' ->> 'id')
WHERE (data -> 'braintreeTransaction' ->> 'id') IS NOT NULL;
```

I won't iterate on that here, mainly because we would need to migrate Stripe transaction's charge ID and the logic using them if we go with a dedicated column, but feel free to propose the change in a dedicated issue. I'm not strongly sided with one of these solutions, but I do agree that having a dedicated column is cleaner.